### PR TITLE
Issue 492 potentiostatic

### LINKED ISSUE #492: potentiostatic mode
Add option to change BCs for setting potential instead of current

### PR DESCRIPTION
**Pull request by tinosulzer**
# Description

Add models for voltage-controlled and power-controlled discharge

Fixes #492 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/master/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: input/parameters/lead-acid/experiments/1C_discharge_from_full/parameters.csv [hunk 1]
```diff
@@ -10,8 +10,7 @@ Number of electrodes connected in parallel to make a cell,8,Manufacturer,
 Number of cells connected in series to make a battery,6,Manufacturer,
 Lower voltage cut-off [V],1.73,,(just under) 10.5V across 6-cell battery
 Upper voltage cut-off [V],2.44,,(just over) 14.5V across 6-cell battery
-C-rate,1,,
-Current function,[constant],,
+C-rate,0.1,,
```
Comment (rtimms) [inline input/parameters/lead-acid/experiments/1C_discharge_from_full/parameters.csv]: should this still be 1?

File: input/parameters/lithium-ion/cells/kokam_Marquis2019/parameters.csv [hunk 2]
```diff
@@ -34,3 +34,4 @@ Positive current collector thermal conductivity [W.m-1.K-1],237,,
 ,,,
 # Electrical,,,
 Cell capacity [A.h],0.680616,,24 Ah/m2 * 0.137m * 0.207m
+Typical current [A],1.46925726107,,1C current
```
Comment (rtimms) [inline input/parameters/lithium-ion/cells/kokam_Marquis2019/parameters.csv]: should this just be equal to the cell capacity?

File: pybamm/models/submodels/external_circuit/function_control_external_circuit.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
+#
+# External circuit with an arbitrary function
+#
+import pybamm
+from .base_external_circuit import BaseModel, LeadingOrderBaseModel
+
+
+class FunctionControl(BaseModel):
```
Comment (rtimms) [inline pybamm/models/submodels/external_circuit/function_control_external_circuit.py]: this is great!

File: pybamm/parameters/parameter_values.py [hunk 4]
```diff
@@ -172,33 +178,33 @@ def update(self, values, check_conflict=False, path=""):
...
                     super().__setitem__(name, value)
+        # check parameter values
+        self.check_and_update_parameter_values(values)
         # reset processed symbols
         self._processed_symbols = {}
 
     def check_and_update_parameter_values(self, values):
-        # Make sure "C-rate" and current are both non-zero
-        if "C-rate" in values and values["C-rate"] == 0:
+        # Make typical current is non-zero
```
Comment (rtimms) [inline pybamm/parameters/parameter_values.py]: make _sure_ typical...

File: input/parameters/lead-acid/experiments/1C_discharge_from_full/parameters.csv (hunk 1 above)
Comment (tinosulzer) [reply input/parameters/lead-acid/experiments/1C_discharge_from_full/parameters.csv]: one of the tests failed when it was 1C so I just lazily changed it so 0.1C, don't really care about lead-acid any more ...

**Pull request merged**